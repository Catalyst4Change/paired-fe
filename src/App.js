import React, { Component } from 'react';
import './styles/main.scss';
import { withRouter, Route, Switch } from 'react-router-dom';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import Sidebar from './containers/Sidebar';

class App extends Component {
  render() {
    return (
      <div className="App">
        <Sidebar />
        <div>
          <header>hi</header>
          <Switch>
            <Route path="/schedule" render={() => <div>path: /schedule</div>} />
            <Route
              path="/avail-pairings"
              render={() => <div>path: /avail-pairings</div>}
            />
            <Route exact path="/" render={() => <div>path: /</div>} />
            <Route render={() => <div>ERRORRRRRRRR</div>} />
          </Switch>
        </div>
      </div>
    );
  }
}

export default withRouter(App);
