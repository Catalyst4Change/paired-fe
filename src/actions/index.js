export const setLoading = bool => ({
  type: 'SET_LOADING',
  bool
});

export const setError = error => ({
  type: 'SET_ERROR',
  error
});

export const setUser = user => ({
  type: 'SET_USER',
  user
});