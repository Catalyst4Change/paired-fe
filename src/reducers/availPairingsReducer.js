export const availPairingsReducer = (state = [], action) => {
  switch (action.type) {
    case 'SET_PAIRINGS':
      return action.pairings;
    case 'SIGN_USER_OUT':
      return [];
    default:
      return state;
  }
}