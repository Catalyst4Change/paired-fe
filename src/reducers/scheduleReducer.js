export const scheduleReducer = (state = [], action) => {
  switch (action.type) {
    case 'SET_SCHEDULE':
      return action.schedule;
    default:
      return state;
  }
};
