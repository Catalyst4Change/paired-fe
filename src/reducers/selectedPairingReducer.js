export const selectedPairingReducer = (state = '', action) => {
  switch (action.type) {
    case 'SET_ID':
      return action.id;
    case 'SIGN_USER_OUT':
      return '';
    default:
      return state;
  }
}