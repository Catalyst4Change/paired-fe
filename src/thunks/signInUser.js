import { fetchData } from '../utils';
import { setError, setLoading, setUser } from '../actions';
import { getSchedule } from './getSchedule';
import * as gql from '../queries';

export const signInUser = (id) => {
  return async dispatch => {
    dispatch(setLoading(true));
    try {
      const userQuery = gql.getUserByFirebaseID(id);
      const { user } = await fetchData(userQuery);
      if (user !== null) {
        dispatch(setUser(user));
        dispatch(getSchedule(user.id));
      } else {
        dispatch(setUser({ isNewUser: true }));
      }
    } catch (error) {
      dispatch(setError(error.message));
    }
    dispatch(setLoading(false));
  };
};
