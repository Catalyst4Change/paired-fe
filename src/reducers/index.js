import { errorReducer } from './errorReducer';
import { loadingReducer } from './loadingReducer';
import { combineReducers } from 'redux';
import { userReducer } from './userReducer';
import { selectedPairingReducer } from './selectedPairingReducer';
import { availPairingsReducer } from './availPairingsReducer';
import { scheduleReducer } from './scheduleReducer';

export const rootReducer = combineReducers({
  isLoading: loadingReducer,
  hasError: errorReducer,
  user: userReducer,
  selectedPairing: selectedPairingReducer,
  availPairings: availPairingsReducer,
  schedule: scheduleReducer
});
