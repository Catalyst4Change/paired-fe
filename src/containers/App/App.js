import React, { Component } from 'react';
import { withRouter, Route, Switch } from 'react-router-dom';
import Sidebar from '../../components/Sidebar';
import { connect } from 'react-redux';
import Controls from '../Controls';
import Schedule from '../Schedule';
import Confirmation from '../Confirmation';
import SignIn from '../SignIn';
import { signUserOut } from '../../actions';
import { signInUser } from '../../thunks/signInUser';
import firebase from 'firebase/app';
import 'firebase/auth';
import codesvg from '../../images/code-typing.svg';

export class App extends Component {
  componentDidMount() {
    this.checkUser();
  }

  checkUser = () => {
    firebase.auth().onAuthStateChanged(async (user) => {
      if (user) {
        this.props.signInUser(user.uid);
      }
    });
  }

  handleSignOut = async () => {
    await firebase.auth().signOut();
    this.props.signUserOut();
  };

  render() {
    const { user } = this.props;
    return (
      <div className='App'>
        <Sidebar />
        <div>
          <header />
          {user.id && (
            <React.Fragment>
              <button
                className='App--button--signout'
                onClick={this.handleSignOut}>
                Sign Out
              </button>
              <Switch>
                <Route path='/schedule' component={Schedule} />
                <Route path='/book-pairing' component={Controls} />
                <Route path='/confirm' component={Confirmation} />
                <Route
                  exact
                  path='/'
                  render={() => (
                    <div>
                      <h2>Welcome {user.name}</h2>
                      <img className='App--img' src={codesvg} alt='Two people coding'/>
                    </div>
                  )}
                />
                <Route
                  render={() => <div>Uh oh! Sorry, page not found.</div>}
                />
              </Switch>
            </React.Fragment>
          )}
          {!user.id && <SignIn history={this.props.history} />}
        </div>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  isLoading: state.isLoading,
  hasError: state.hasError,
  user: state.user
});

export const mapDispatchToProps = dispatch => ({
  signInUser: (id) => dispatch(signInUser(id)),
  signUserOut: () => dispatch(signUserOut())
});

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(App)
);
