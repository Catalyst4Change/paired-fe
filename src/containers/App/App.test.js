import React from 'react';
import ReactDOM from 'react-dom';
import { App, mapStateToProps, mapDispatchToProps} from './App';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import { rootReducer } from '../../reducers';
import * as actions from '../../actions';
import { signInUser } from '../../thunks/signInUser';
import { BrowserRouter } from 'react-router-dom';
import firebase from 'firebase/app';

jest.mock('../../thunks/signInUser');

beforeAll(() => {
  firebase.initializeApp({
    apiKey: 'AIzaSyBZztEJm_KxmHfMGaxYgLIFah2XAdS5mh0',
    authDomain: 'paired-turing.firebaseapp.com'
  });
});

describe('App', () => {
  let wrapper;
  const mockProps = {
    hasError: '',
    isLoading: false,
    user: { id: 'abc123' },
    signInUser: jest.fn(),
    signUserOut: jest.fn(),
    history: {
      push: jest.fn()
    }
  };

  beforeEach(() => {
    wrapper = shallow(<App {...mockProps} />);
  });

  describe('render', () => {
    it('renders without crashing', () => {
      const div = document.createElement('div');
      const store = createStore(rootReducer);
      ReactDOM.render(
        <BrowserRouter>
          <Provider store={store}>
            <App {...mockProps} />
          </Provider>
        </BrowserRouter>,
        div
      );
      ReactDOM.unmountComponentAtNode(div);
    });

    it('should match the snapshot when there is a user id', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should match the snapshot when there is no user id', () => {
      wrapper = shallow(<App {...mockProps} user={{}}/>);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('componentDidMount', () => {
    it('should call checkUser', () => {
      jest.spyOn(wrapper.instance(), 'checkUser');
      wrapper.instance().forceUpdate();
      wrapper.instance().componentDidMount();
      expect(wrapper.instance().checkUser).toHaveBeenCalled();
    });
  });

  describe('handleSignOut', () => {
    it('should call signUserOut', async () => {
      await wrapper.instance().handleSignOut();
      expect(mockProps.signUserOut).toHaveBeenCalled();
    });
  });
});

describe('mapStateToProps', () => {
  it('should return a props object with the correct properties', () => {
    const initialState = {
      isLoading: false,
      hasError: false,
      user: { id: 'abc123' },
      extraProperty: true
    };
    const expected = {
      isLoading: false,
      hasError: false,
      user: { id: 'abc123' }
    };
    const result = mapStateToProps(initialState);
    expect(result).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {
  const mockDispatch = jest.fn();
  const mappedProps = mapDispatchToProps(mockDispatch);

  it('should dispatch signInUser', () => {
    const expected = signInUser('abc123')
    mappedProps.signInUser('abc123');
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
  
  it('should dispatch signUserOut', () => {
    const expected = actions.signUserOut();
    mappedProps.signUserOut();
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
});
