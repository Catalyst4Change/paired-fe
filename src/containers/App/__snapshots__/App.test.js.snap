// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render should match the snapshot when there is a user id 1`] = `
<div
  className="App"
>
  <Header
    handleSignOut={[Function]}
    user={
      Object {
        "id": "abc123",
      }
    }
  />
  <div>
    <Switch>
      <Route
        component={[Function]}
        path="/schedule"
      />
      <Route
        component={[Function]}
        path="/book-pairing"
      />
      <Route
        component={[Function]}
        path="/confirm"
      />
      <Route
        component={[Function]}
        path="/edit-availability"
      />
      <Route
        exact={true}
        path="/"
        render={[Function]}
      />
      <Route
        render={[Function]}
      />
    </Switch>
  </div>
</div>
`;

exports[`App render should match the snapshot when there is no user id 1`] = `
<div
  className="App"
>
  <Header
    handleSignOut={[Function]}
    user={Object {}}
  />
  <div>
    <Switch>
      <Route
        exact={true}
        path="/"
        render={[Function]}
      />
      <Route
        path="/set-availability"
        render={[Function]}
      />
      <Route
        path="/schedule"
        render={[Function]}
      />
      <Route
        path="/book-pairing"
        render={[Function]}
      />
      <Route
        render={[Function]}
      />
    </Switch>
  </div>
</div>
`;
